File: packages/companion/swift/Sources/BridgethingCompanion/TransferPacer.swift
import Foundation

struct TransferPacer {
    static let targetDelaySeconds: Double = 0.6
    static let ackIntervalBytes: UInt64 = 16 * 1024
    static let minWindowBytes: UInt64 = 4 * TransferPacer.ackIntervalBytes
    static let maxWindowBytes: UInt64 = 16 * TransferPacer.ackIntervalBytes
    static let fragmentBytes: Int = 16 * 1024
    static let rateSampleCount: Int = 8

    private let clock: () -> Double
    private var ackedBytes: UInt64
    private var lastProgressAt: Double
    private var samples: [Double] = []

    init(startOffset: UInt64 = 0, clock: @escaping () -> Double = { ProcessInfo.processInfo.systemUptime }) {
        self.clock = clock
        ackedBytes = startOffset
        lastProgressAt = clock()
    }

    var ratePerSec: Double? { samples.max() }

    var windowBytes: UInt64 {
        guard let rate = ratePerSec else { return Self.minWindowBytes }
        let budget = UInt64((rate * Self.targetDelaySeconds).rounded())
        return min(max(budget, Self.minWindowBytes), Self.maxWindowBytes)
    }

    var fragmentBytes: Int { Self.fragmentBytes }

    mutating func observe(ackedBytes acked: UInt64) {
        guard acked > ackedBytes else { return }
        let now = clock()
        let dt = max(now - lastProgressAt, 0.001)
        samples.append(Double(acked - ackedBytes) / dt)
        if samples.count > Self.rateSampleCount { samples.removeFirst(samples.count - Self.rateSampleCount) }
        ackedBytes = acked
        lastProgressAt = now
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/TransferPacer.kt
package com.bridgething.companion

internal class TransferPacer(
    startOffset: Long = 0L,
    private val clock: () -> Double = { System.nanoTime() / 1_000_000_000.0 },
) {
    private var ackedBytes: Long = startOffset
    private var lastProgressAt: Double = clock()
    private val samples = ArrayDeque<Double>()

    val ratePerSec: Double?
        get() = samples.maxOrNull()

    val windowBytes: Long
        get() {
            val rate = ratePerSec ?: return MIN_WINDOW_BYTES
            val budget = (rate * TARGET_DELAY_SECONDS).toLong()
            return budget.coerceIn(MIN_WINDOW_BYTES, MAX_WINDOW_BYTES)
        }

    val fragmentBytes: Int get() = FRAGMENT_BYTES

    fun observe(acked: Long) {
        if (acked <= ackedBytes) return
        val now = clock()
        val dt = (now - lastProgressAt).coerceAtLeast(0.001)
        samples.addLast((acked - ackedBytes) / dt)
        while (samples.size > RATE_SAMPLE_COUNT) samples.removeFirst()
        ackedBytes = acked
        lastProgressAt = now
    }

    internal companion object {
        const val TARGET_DELAY_SECONDS: Double = 0.6

        const val ACK_INTERVAL_BYTES: Long = 16 * 1024L
        const val MIN_WINDOW_BYTES: Long = 4 * ACK_INTERVAL_BYTES

        const val MAX_WINDOW_BYTES: Long = 16 * ACK_INTERVAL_BYTES
        const val FRAGMENT_BYTES: Int = 16 * 1024
        const val RATE_SAMPLE_COUNT: Int = 8
    }
}
